File: 백준/Gold/28703. Double It/Double It.py
import sys
import heapq

N = int(sys.stdin.readline()[:-1])
A = []
mx = 0

for v in map(int, sys.stdin.readline()[:-1].split()):
    mx = max(mx, v)
    heapq.heappush(A, v)

curr_mx = mx
diff = curr_mx - A[0] #현재 최댓값과 최솟값의 차이

while A[0] < mx:
    v = heapq.heappop(A)
    diff = min(diff, curr_mx - v)
    curr_mx = max(curr_mx, 2*v)
    heapq.heappush(A, 2*v)
print(min(curr_mx-A[0], diff))

File: 백준/Gold/28703. Double It/Double It.cc
#include <iostream>
#include <vector>
#include <queue>
using namespace std;


int main() {
    int N; cin >> N;

    priority_queue<int, vector<int>, greater<int>> Q;
    int mx = 0;
    for (int i=0; i<N; i++){
        int v; cin >> v;
        mx = max(mx, v);
        Q.push(v);
    }

    int curr_mx = mx;
    int diff = curr_mx - Q.top();

    while (Q.top() < mx){
        int v = Q.top(); Q.pop();
        diff = min(diff, curr_mx - v);
        curr_mx = max(curr_mx, v*2);
        Q.push(v*2);
    }

    cout << min(curr_mx - Q.top(), diff);

    return 0;
}